### PR TITLE
[#120455] Verify splits have unique subaccounts

### PR DESCRIPTION
**Pull request by benlinton**
Test that you cannot have duplicate accounts in a split

https://www.pivotaltracker.com/story/show/114612951

### CODE REVIEW

File: vendor/engines/split_accounts/spec/models/split_account_spec.rb [hunk 1]
```diff
@@ -58,8 +61,23 @@
...
-          split_account.splits << build(:split, percent: 50, extra_penny: true, parent_split_account: split_account)
-          split_account.splits << build(:split, percent: 50, extra_penny: true, parent_split_account: split_account)
+          split_account.splits << build(:split, percent: 50, extra_penny: true, parent_split_account: split_account, subaccount: subaccount_1)
+          split_account.splits << build(:split, percent: 50, extra_penny: true, parent_split_account: split_account, subaccount: subaccount_2)
+        end
+      end
+
+      it "is invalid" do
+        expect(split_account).not_to be_valid
+        expect(split_account.errors).to include(:splits)
```
Comment (jhanggi) [inline vendor/engines/split_accounts/spec/models/split_account_spec.rb]: I feel like now that we're adding more validations, we should probably check that it's invalid for the right reason.

File: vendor/engines/split_accounts/app/models/split_accounts/split_account.rb [hunk 2]
```diff
@@ -25,10 +26,23 @@ def one_split_has_extra_penny
...
+        errors.add(:splits, :duplicate_subaccounts)
+      end
+    end
+
     def extra_penny_count
       splits.select(&:extra_penny?).size
     end
 
+    def duplicate_subaccounts?
+      splits.map(&:subaccount_id)
```
Comment (jhanggi) [inline vendor/engines/split_accounts/app/models/split_accounts/split_account.rb]: It's not that bad, but it seems like there should be a cleaner way to write this.

```
subaccounts = splits.map(&:subaccount_id)
subaccounts.length == subaccounts.uniq.length
```
or
```
splits.map(&:subaccount_id)
  .group_by { |e| e }
  .any? { |k, v| v.size > 1 }
```

File: vendor/engines/split_accounts/spec/models/split_account_spec.rb (hunk 1 above)
Comment (benlinton) [inline vendor/engines/split_accounts/spec/models/split_account_spec.rb]: Done.

File: vendor/engines/split_accounts/app/models/split_accounts/split_account.rb (hunk 2 above)
Comment (benlinton) [inline vendor/engines/split_accounts/app/models/split_accounts/split_account.rb]: Done.  I believe the second option is supposed to be much more performant.

**Comment by benlinton** [comment]
@jhanggi Ready for a re-review.

File: vendor/engines/split_accounts/spec/models/split_account_spec.rb [hunk 3]
```diff
@@ -30,42 +33,61 @@
...
-          split_account.splits << build(:split, percent: 50, extra_penny: false, parent_split_account: split_account)
+          split_account.splits << build(:split, percent: 50, extra_penny: false, parent_split_account: split_account, subaccount: subaccount_1)
+          split_account.splits << build(:split, percent: 50, extra_penny: false, parent_split_account: split_account, subaccount: subaccount_2)
         end
       end
 
       it "is invalid" do
         expect(split_account).not_to be_valid
-        expect(split_account.errors).to include(:splits)
+        message = I18n.t("activerecord.errors.models.split_accounts/split_account.attributes.splits.only_one_extra_penny")
```
Comment (jhanggi) [inline vendor/engines/split_accounts/spec/models/split_account_spec.rb]: Seems like there should be a cleaner way to do this test, but I haven't seen a better way.

**Comment by jhanggi** [comment]
:+1:

**Pull request closed without merging**